reader_concurrency_semaphore.cc

```diff
@@ -149,14 +152,33 @@ bool reader_concurrency_semaphore::try_evict_one_inactive_read() {
     return true;
 }
 
+class queue_throttler {
+    static constexpr size_t max_error_response_delay_ms = 1000;
+    size_t _severity;
+public:
+    queue_throttler(size_t current, size_t max)
+            : _severity(current > max ? current - max : 0) {}
+    bool should_throttle() {
+        static thread_local std::default_random_engine random_engine{std::random_device{}()};
+        static thread_local std::uniform_int_distribution<size_t> random_dist{0, 99};
+        return _severity > 0 && (_severity >= 100 || random_dist(random_engine) < _severity);
+    }
+    future<reader_permit::resource_units> throttle(sstring queue_name) {
+        return seastar::sleep(std::chrono::milliseconds(std::min(10 * _severity, max_error_response_delay_ms))).then([queue_name = std::move(queue_name)] {
+            return make_exception_future<reader_permit::resource_units>(
+                    std::make_exception_ptr(std::runtime_error(
+                            format("{}: restricted mutation reader queue overload", queue_name))));
+        });
```

Comment:
I both agree and disagree with this:

 - disagree: we are delaying the client from knowing that we will not be serving it, and so tying up client resources that could be well spent doing other things.
 - agree: it's not realistic to expect we'll be able to change all the client drivers, and get users to upgrade to using the updated drivers

On the whole I agree more than I disagree. A way to resolve the conflict is to allow the user to declare they are backing off on the client side, with some switch, and disable the mechanism.

Reply:
One more problem is that in this particular place we also tie up coordinator's resources. On the other hand, this error should be treated as last hope before going OOM, after all other coordinator-side load shedding efforts failed.